components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,343 @@
+        "
+        SELECT
+          INSTR(:query, ' ' || keyword || ' ') AS sign_index,
+          LENGTH(keyword) AS sign_length,
+          need_location
+        FROM yelp_location_signs
+        WHERE sign_index ORDER BY sign_length DESC LIMIT 1
+        ",
+        named_params! {
+            ":query": &query_with_spaces.to_lowercase(),
+        },
+        |row| -> Result<_> {
+            Ok((
+                row.get::<_, usize>("sign_index")?,
+                row.get::<_, usize>("sign_length")?,
+                row.get::<_, i8>("need_location")?,
+            ))
+        },
+    )?;
+
+    if results.is_empty() {
+        Ok((query.trim().to_string(), None, None, false))
+    } else {
+        let (sign_index, sign_length, need_location) = &results[0];
+        let pre_location = query_with_spaces.get(..*sign_index).unwrap().trim();
+        let location_sign = query_with_spaces
+            .get(*sign_index..*sign_index + *sign_length)
+            .unwrap()
+            .trim();
+        let location = query_with_spaces
+            .get(*sign_index + *sign_length..)
+            .unwrap()
+            .trim();
+        Ok((
+            pre_location.to_string(),
+            if location_sign.is_empty() {
+                None
+            } else {
+                Some(location_sign.to_string())
+            },
+            if location.is_empty() {
+                None
+            } else {
+                Some(location.to_string())
+            },
+            *need_location == 1,
+        ))
+    }
+}
+
+fn find_modifiers(
+    conn: &Connection,
+    query: &str,
+) -> Result<(String, Option<String>, Option<String>)> {
+    if !query.contains(" ") {
+        return Ok((query.to_string(), None, None));
+    }
+
+    let words: Vec<_> = query.split_whitespace().collect();
+
+    let mut pre_modifier = None;
+    for i in 0..MAX_MODIFIER_WORDS_LENGTH {
+        let mut candidate_chunks = words.chunks(MAX_MODIFIER_WORDS_LENGTH - i);
+        let candidate = candidate_chunks.next().unwrap().join(" ");
+        // Search from longer term.
+        if is_modifier(conn, &candidate, Modifier::PreModifier)? {
+            pre_modifier = Some(candidate);
+            break;
+        }
+    }
+
+    let mut post_modifier = None;
+    for i in 0..MAX_MODIFIER_WORDS_LENGTH {
+        let mut candidate_chunks = words.rchunks(MAX_MODIFIER_WORDS_LENGTH - i);
+        let candidate = candidate_chunks.next().unwrap().join(" ");
+        if is_modifier(conn, &candidate, Modifier::PostModifier)? {
+            post_modifier = Some(candidate);
+            break;
+        }
+    }
+
+    let mut subject_candidate = query;
+    if let Some(ref modifier) = pre_modifier {
+        subject_candidate = &subject_candidate[modifier.len()..];
+    }
+    if let Some(ref modifier) = post_modifier {
+        subject_candidate = &subject_candidate[..subject_candidate.len() - modifier.len()];
```

Comment:
In the special case that there is only a pre-modifier and a post-modifier but not the subject, e.g. `best delivery`, would it only return a empty string as the subject?

Reply:
Another special case: if both the pre and the post modifiers are the same, and the `query` is exactly the modifier, this will crash because `subject_candidate.len() - modifier.len()` would overflow. Let's be extra cautious when manipulating those indices.

Reply:
> In the special case that there is only a pre-modifier and a post-modifier but not the subject, e.g. `best delivery`, would it only return a empty string as the subject?

Yes, then, as we check it by `is_subject()`, it should be no problem, I think.
But, the `is_subject()` does not check whether the query is empty, I will add the check.
Thanks!

> Another special case: if both the pre and the post modifiers are the same, and the `query` is exactly the modifier, this will crash because `subject_candidate.len() - modifier.len()` would overflow. Let's be extra cautious when manipulating those indices.

If the query was one word such as the `best`, it was possible to touch the failure.
But, we check whether the query has a whitespace at the beginning, the failure never happens, I think.
https://github.com/mozilla/application-services/pull/5951/files#diff-22307d3249667224ac6502dabf1e4b86e075333c0f279d93b126d25a8d7bb7c8R119
To be safe, I add the tests.